src/api/app/models/user.rb

```diff
@@ -80,36 +78,20 @@ class User < ApplicationRecord
                       message:     'must be a valid email address.',
                       allow_blank: true }
 
-  # We want to validate the format of the password and only allow alphanumeric
-  # and some punctiation/base64 characters.
-  # The format must only be checked if the password has been set and the record
-  # has not been stored yet.
+  # only allow alphanumeric and some punctiation/base64 characters
   validates :password,
             format: { with:    %r{\A[\w\.\- /+=!?(){}|~*]+\z},
-                      message: 'must not contain invalid characters.',
-                      if:      Proc.new { |user| user.password_changed? && !user.password_was.nil? } }
+                      message: 'must not contain invalid characters.' },
+            allow_nil: true
```

Comment:
Why is the allow_nil necessary?

Reply:
I guess because you want to set the old_password to nil in the conver_password? However, then you need to rename the validation to :password_old?

Reply:
Nope. `password` is the correct name. It's a hidden instance variable which is set when using `has_secure_password`. Any value you assign to it will be stored as a BCrypt string in `password_digest`. It's just to make it easier for the developer to change passwords instead of going the way over the `bcrypt` library.

The `allow_nil` is also necessary since `password` is always `nil` because it's not a database structure field. It's just a kind of a wrapper for `password_digest` as described above.

Reply:
Then I suggest dropping the whole validation as I don't see the point in limit the allowed characters.
https://blog.codinghorror.com/password-rules-are-bullshit/

Regarding the allow_nil: This is not correct as ``password=`` will set the ``password`` attribute, so it shouldn't be nil anymore. As explanation have a look how [ActiveModel::AttributeMethods](http://api.rubyonrails.org/classes/ActiveModel/AttributeMethods.html) works.

Here is the implementation of ``has_secure_password``
https://github.com/rails/rails/blob/18a2513729ab90b04b1c86963e7d5b9213270c81/activemodel/lib/active_model/secure_password.rb#L116

Reply:
Regarding the allow_nil:
```
irb(main):013:0> u = User.first
  User Load (0.9ms)  SELECT  `users`.* FROM `users` ORDER BY `users`.`id` ASC LIMIT 1
=> #<User id: 1, created_at: "2017-09-13 09:45:09", updated_at: "2017-09-13 10:18:17", last_logged_in_at: "2017-09-13 10:17:07", login_failure_count: 0, login: "Admin", email: "root@localhost", password_digest: "$2a$10$de2vLn6PgJb4d33EfSFaOO.eg4GhlZtuBMOfM0Ae8T6...", realname: "OBS Instance Superuser", password_old: nil, password_old_hash_type: nil, password_old_salt: nil, adminnote: nil, state: "confirmed", owner_id: nil>
irb(main):014:0> u.password
=> nil
irb(main):015:0> u.valid?
  User Exists (0.7ms)  SELECT  1 AS one FROM `users` WHERE `users`.`login` = 'Admin' AND (`users`.`id` != 1) LIMIT 1
=> false
irb(main):016:0> u.errors.full_messages.to_sentence
=> "Password must not contain invalid characters. and Password must have between 6 and 64 characters."
irb(main):017:0> 
```

It would be VERY(!) strange if it would save the not-hashed password in an instance variable... As I said before, this attribute is only being used for setting/changing the password and is always `nil` if you load a user from the database as you can see above.

______

Anyways, I'm fine with dropping the "invalid characters" validation.

Reply:
Yeah, right, it is only set when you use the ``password=`` method not when you load from the db. Yes, please just drop the validations.